pkg/kubectl/cmd/apply.go

```diff
@@ -154,43 +158,16 @@ func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *Ap
 			return nil
 		}
 
-		// Serialize the current configuration of the object from the server.
-		current, err := runtime.Encode(encoder, info.Object)
-		if err != nil {
-			return cmdutil.AddSourceToErr(fmt.Sprintf("serializing current configuration from:\n%v\nfor:", info), info.Source, err)
-		}
-
-		// Retrieve the original configuration of the object from the annotation.
-		original, err := kubectl.GetOriginalConfiguration(info)
-		if err != nil {
-			return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving original configuration from:\n%v\nfor:", info), info.Source, err)
-		}
-
-		// Create the versioned struct from the original from the server for
-		// strategic patch.
-		// TODO: Move all structs in apply to use raw data. Can be done once
-		// builder has a RawResult method which delivers raw data instead of
-		// internal objects.
-		versionedObject, _, err := decoder.Decode(current, nil, nil)
-		if err != nil {
-			return cmdutil.AddSourceToErr(fmt.Sprintf("converting encoded server-side object back to versioned struct:\n%v\nfor:", info), info.Source, err)
-		}
-
-		// Compute a three way strategic merge patch to send to server.
-		patch, err := strategicpatch.CreateThreeWayMergePatch(original, modified, current, versionedObject, true)
-		if err != nil {
-			format := "creating patch with:\noriginal:\n%s\nmodified:\n%s\ncurrent:\n%s\nfrom:\n%v\nfor:"
-			return cmdutil.AddSourceToErr(fmt.Sprintf(format, original, modified, current, info), info.Source, err)
-		}
-
 		helper := resource.NewHelper(info.Client, info.Mapping)
-		_, err = helper.Patch(info.Namespace, info.Name, api.StrategicMergePatchType, patch)
+		patcher := &patcher{encoder, decoder, info.Mapping, helper}
+
+		err = patchResource(patcher, info.Object, modified, info.Source, info.Namespace, info.Name)
 		if err != nil {
-			return cmdutil.AddSourceToErr(fmt.Sprintf("applying patch:\n%s\nto:\n%v\nfor:", patch, info), info.Source, err)
+			return cmdutil.AddSourceToErr(fmt.Sprintf("applying patch to:\n%v\nfor:", info), info.Source, err)
 		}
 
 		if cmdutil.ShouldRecord(cmd, info) {
-			patch, err = cmdutil.ChangeResourcePatch(info, f.Command())
+			patch, err := cmdutil.ChangeResourcePatch(info, f.Command())
```

Comment:
We should at least add a TODO here to figure out if it should also be retried.  Since it isn't updated atomically with the changes to the original object, there may be some complications figuring out what to do if there is a conflict at this point.